nominal/core/stream.py

```diff
@@ -99,25 +108,31 @@ def flush(self, wait: bool = False, timeout: float | None = None) -> None:
                      NOTE: If none, waits indefinitely.
 
         """
-        if not self._batch:
-            logger.debug("Not flushing... no enqueued batch")
-            return
+        with self._batch_lock:
+            if not self._batch:
+                logger.debug("Not flushing... no enqueued batch")
+                self._last_batch_time = time.time()
+                return
+
+        self._pending_jobs.acquire()
 
         def process_future(fut: concurrent.futures.Future) -> None:  # type: ignore[type-arg]
             """Callback to print errors to the console if a batch upload fails."""
+            self._pending_jobs.release()
             maybe_ex = fut.exception()
             if maybe_ex is not None:
                 logger.error("Batched upload task failed with exception", exc_info=maybe_ex)
             else:
                 logger.debug("Batched upload task succeeded")
 
-        logger.debug(f"Starting flush with {len(self._batch)} records")
-        future = self._executor.submit(self._process_batch, self._batch)
-        future.add_done_callback(process_future)
+        with self._batch_lock:
+            logger.debug(f"Starting flush with {len(self._batch)} records")
+            future = self._executor.submit(self._process_batch, self._batch)
+            future.add_done_callback(process_future)
 
-        # Clear metadata
-        self._batch = []
-        self._last_batch_time = time.time()
+            # Clear metadata
+            self._batch = []
+            self._last_batch_time = time.time()
```

Comment:
its not obvious which variables need to hold the lock. Wonder if this could become more clear using a queue or similar structure

Reply:
I've refactored a bit so it's a bit more clear. We also need the lock to coordinate access _last_batch_time, I don't think a queue would solve that.